no-pdi/include/pdi.h

```diff
@@ -0,0 +1,288 @@
+* THE SOFTWARE.
+******************************************************************************/
+
+/** \file pdi.h for deactivation
+ *
+ * C user API
+ *
+ * The user facing API is the interface offered by PDI to C application
+ * developers.
+ *
+ * \defgroup init_final Initialization and finalization
+ *
+ * The initialization and finalization part of the API is used to setup PDI,
+ * release its resources and check version information.
+ *
+ * \defgroup annotation Code annotation
+ *
+ * The code annotation API is the main interface to use in the code.
+ *
+ * It offers functions that can be called from code with no side effect by
+ * default and that can therefore be considered as annotations.
+ *
+ * \defgroup error Error handling
+ *
+ * The error handling API supports checking the error status of PDI.
+ *
+ * By default, errors in PDI C API are signaled by a return code of type
+ * PDI_status_t and an error message can be retrieved with the PDI_errmsg
+ * function. This default behavior can be changed by replacing the error handler
+ * with the PDI_errhandler function.
+ *
+ */
+
+#ifndef PDI_H_
+#define PDI_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** \addtogroup error
+ * \{
+ */
+
+/** Error codes of PDI
+ */
+typedef enum PDI_status_e {
+	/// everything went well
+	PDI_OK = 0,
+	/// on an input call, no such data is available
+	PDI_UNAVAILABLE,
+	/// The configuration file is invalid
+	PDI_ERR_CONFIG,
+	/// A value expression is invalid
+	PDI_ERR_VALUE,
+	/// Tried to load a non-existing plugin
+	PDI_ERR_PLUGIN,
+	/// Implementation limitation (typically an unimplemented feature)
+	PDI_ERR_IMPL,
+	/// A system error occured (OS, etc.)
+	PDI_ERR_SYSTEM,
+	/** A call to a function has been made at a wrong time (e.g. closing an
+	 *  unopened transaction)
+	 */
+	PDI_ERR_STATE,
+	/// A conflict of onwership over a content has been raised
+	PDI_ERR_RIGHT,
+	/// Invalid type error
+	PDI_ERR_TYPE
+
+} PDI_status_t;
+
+/** Type of a callback function used when an error occurs
+ * \param status the error code
+ * \param message the human-readable error message
+ * \param context a user-provided context
+ */
+typedef void (*PDI_errfunc_f)(PDI_status_t status, const char* message, void* context);
+
+/** Definition of an error handler
+ */
+typedef struct PDI_errhandler_s {
+	/// The function to handle the error (none if NULL)
+	PDI_errfunc_f func;
+
+	/// the context that will be provided to the function
+	void* context;
+
+} PDI_errhandler_t;
+
+/** Prints the error message and aborts if the status is invalid
+ */
+extern const PDI_errhandler_t PDI_ASSERT_HANDLER;
+
+/** Prints the error message and continue if the status is invalid
+ */
+extern const PDI_errhandler_t PDI_WARN_HANDLER;
+
+/** Does nothing
+ */
+const PDI_errhandler_t PDI_NULL_HANDLER = {NULL, NULL};
+
+/** Return a human-readabe message describing the last error that occured in PDI
+ */
+static inline const char* PDI_errmsg(void)
+{
+	return 0;
```

Comment:
Maybe to be on the safe side:
```suggestion
	return "";
```

Reply:
If the result of the function is passed to printf, a null pointer will make it crash

Reply:
Checking for equality of output may result in unstable behaviour.
`comparison with string literal results in unspecified behavior [-Waddress]`

Reply:
indeed, in this case you should not compare with `==` but with `strcmp` or something equivalent.